File: retrocpu_test_framework/src/mn1613/cpu_log.ts
/**
 * テスト専用 CPU 実行ログ
 * 根拠: asm_test_framework.mdc §テスト専用 CPU ログ出力
 */

import fs from "node:fs";
import path from "node:path";
import {
  getClockCount,
  type CPURegister,
} from "../../../retrocpu_emu/src/cpuboard/mn1613/mn1613.js";
import { parseCdb } from "../../../retrocpu_emu/src/code_test/cdb.js";
import { Mn1613Disassembler } from "../../../retrocpu_emu/src/dis_assembler/mn1613/index.js";
import {
  setActiveCpuLogMarker,
  takePendingCpuLogTestName,
  type CpuLogTestPhase,
} from "../cpu_log_mark.js";
import type { CpuLogMode } from "../types.js";

const STACK_WORDS = 16;

/**
 * 16bit を 4 桁大文字 16 進にする。
 * @param n 値
 * @returns 例 "0200"
 */
function hex4(n: number): string {
  return (n & 0xffff).toString(16).toUpperCase().padStart(4, "0");
}

/**
 * CDB のチェックポイントからワードアドレス → ログ用名を作る。
 * 同一ワードに複数あれば `,` で連結する（`add_enter$0001,add_leave$0001`）。
 * @param cdbText CDB 全文
 * @returns ワードアドレス → `name$serial`
 */
export function checkpointsByWordAddr(cdbText: string): Map<number, string> {
  const map = new Map<number, string>();
  for (const cp of parseCdb(cdbText).checkpoints) {
    const label = `${cp.name}$${cp.serial}`;
    const addr = cp.wordAddr & 0xffff;
    const prev = map.get(addr);
    map.set(addr, prev ? `${prev},${label}` : label);
  }
  return map;
}

/**
 * レジスタをログ用 1 フィールドにする。
 * @param st CPU 状態（フェッチ直前）
 * @returns 空白区切り
 */
export function formatCpuLogRegs(st: CPURegister): string {
  return [
    `R0=${hex4(st.R[0]!)}`,
    `R1=${hex4(st.R[1]!)}`,
    `R2=${hex4(st.R[2]!)}`,
    `R3=${hex4(st.R[3]!)}`,
    `R4=${hex4(st.R[4]!)}`,
    `SP=${hex4(st.SP)}`,
    `STR=${hex4(st.STR)}`,
    `IC=${hex4(st.IC)}`,
    `CSBR=${(st.CSBR & 0xf).toString(16).toUpperCase()}`,
    `SSBR=${(st.SSBR & 0xf).toString(16).toUpperCase()}`,
    `TSR0=${(st.TSR0 & 0xf).toString(16).toUpperCase()}`,
    `TSR1=${(st.TSR1 & 0xf).toString(16).toUpperCase()}`,
    `OSR0=${hex4(st.OSR[0]!)}`,
    `OSR1=${hex4(st.OSR[1]!)}`,
    `OSR2=${hex4(st.OSR[2]!)}`,
    `OSR3=${hex4(st.OSR[3]!)}`,
    `NPP=${(st.NPP & 0xff).toString(16).toUpperCase().padStart(2, "0")}`,
    `IISR=${hex4(st.IISR)}`,
    `SBRB=${hex4(st.SBRB)}`,
  ].join(" ");
}

/**
 * SP+1 から 16 ワードを読む（空きスロットの次＝スタック先頭）。
 * @param sp 空きスロット
 * @param readWord ワードアドレス → 16bit
 * @returns 空白区切り hex4
 */
export function formatCpuLogStack(
  sp: number,
  readWord: (wordAddr: number) => number,
): string {
  const words: string[] = [];
  for (let i = 1; i <= STACK_WORDS; i += 1) {
    words.push(hex4(readWord((sp + i) & 0xffff)));
  }
  return words.join(" ");
}

/** チェックポイント命令の実行前／実行後（instruction モードは after のみ） */
export type CpuLogPhase = "before" | "after";

type PendingInstruction = {
  ic: number;
  /** チェックポイント名。無ければ `-` */
  name: string;
  hit: number;
  data: string;
  disasm: string;
};

/**
 * テスト専用 CPU ログ。
 * 本文モード無し（`null`）は START/END のみ。本文は TAB 区切り:
 * clock / addr / data / checkpoint / phase / hits / disasm / regs / stack16
 */
export class CpuExecutionLog {
  private readonly filePath: string;
  private mode: CpuLogMode | null;
  private readonly checkpoints: Map<number, string>;
  private readonly disasm: Mn1613Disassembler;
  private readonly readWord: (wordAddr: number) => number;
  private readonly hits = new Map<number, number>();
  private pending: PendingInstruction | null = null;

  /**
   * @param filePath 出力ファイル（既存なら切り詰める）
   * @param cdbText ラベル／チェックポイント用 CDB
   * @param readWord ワード読み（論理アドレス、CSBR=0 前提）
   * @param mode 本文モード。`null` / 省略は START/END のみ
   */
  constructor(
    filePath: string,
    cdbText: string,
    readWord: (wordAddr: number) => number,
    mode: CpuLogMode | null = null,
  ) {
    this.filePath = path.resolve(filePath);
    this.mode = mode;
    this.checkpoints = checkpointsByWordAddr(cdbText);
    this.disasm = new Mn1613Disassembler({ cdbText });
    this.readWord = readWord;
    fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
    fs.writeFileSync(this.filePath, "", "utf8");
    setActiveCpuLogMarker(this);
    const pending = takePendingCpuLogTestName();
    if (pending) {
      this.appendTestMark(pending, "START");
    }
  }

  /** 現在の本文モード。`null` は START/END のみ */
  get logMode(): CpuLogMode | null {
    return this.mode;
  }

  /**
   * 出力モードを切り替える（同一ファイル・同一セッション）。
   * @param mode `checkpoint` / `instruction`。`null` で START/END のみ
   */
  setMode(mode: CpuLogMode | null): void {
    this.mode = mode;
    this.pending = null;
  }

  /**
   * `test()` タイトルを START / END で追記する。
   * @param name ケース名
   * @param phase START / END
   */
  appendTestMark(name: string, phase: CpuLogTestPhase): void {
    fs.appendFileSync(this.filePath, `${name} ${phase}\n`, "utf8");
  }

  /** このテスト開始時点からの通過回数を捨てる（reload 時）。 */
  resetHits(): void {
    this.hits.clear();
    this.pending = null;
  }

  /**
   * セクション見出しを書く（本文モード無しでは書かない）。
   * @param kind reload / runInit / call
   * @param label 対象ラベル（無ければ空）
   */
  beginRun(kind: string, label = ""): void {
    this.pending = null;
    if (this.mode === null) {
      return;
    }
    const extra = label ? ` ${label}` : "";
    fs.appendFileSync(this.filePath, `# ${kind}${extra}\n`, "utf8");
  }

  /**
   * フェッチ直前。
   * 本文無し: 何もしない。
   * `checkpoint`: チェックポイント命令のときだけ before を書く。
   * `instruction`: 命令情報だけ保持（行は出さない）。
   * @param state IRQ 処理後・フェッチ前
   */
  onBeforeExecute(state: CPURegister): void {
    if (this.mode === null) {
      this.pending = null;
      return;
    }

    const ic = state.IC & 0xffff;
    const cpName = this.checkpoints.get(ic);

    if (this.mode === "checkpoint") {
      if (!cpName) {
        this.pending = null;
        return;
      }
      const pending = this.capturePending(ic, cpName);
      this.pending = pending;
      this.writeRecord("before", pending, state);
      return;
    }

    const pending = this.capturePending(ic, cpName ?? "-");
    this.pending = pending;
  }

  /**
   * 命令実行直後。
   * 本文無し: 何もしない。
   * `checkpoint`: 直前がチェックポイントなら after を書く。
   * `instruction`: 直前に保持した命令の after を書く。
   * @param state 実行後（IC は次命令／分岐先のことがある）
   */
  onAfterExecute(state: CPURegister): void {
    if (this.mode === null) {
      this.pending = null;
      return;
    }
    const pending = this.pending;
    if (!pending) return;
    this.pending = null;
    this.writeRecord("after", pending, state);
  }

  /**
   * 実行前の命令情報を撮る（ヒット数を増やす）。
   * @param ic 命令ワードアドレス
   * @param name チェックポイント名または `-`
   * @returns pending
   */
  private capturePending(ic: number, name: string): PendingInstruction {
    const hit = (this.hits.get(ic) ?? 0) + 1;
    this.hits.set(ic, hit);
    const dis = this.disasm.disassemble(ic, this.readWord);
    const dataWords: string[] = [];
    for (let i = 0; i < dis.wordCount; i += 1) {
      dataWords.push(hex4(this.readWord((ic + i) & 0xffff)));
    }
    return {
      ic,
      name,
      hit,
      data: dataWords.join(" "),
      disasm: dis.text,
    };
  }

  /**
   * 1 行追記する。addr / data / disasm は実行した命令側。
   * @param phase before / after
   * @param pending 実行前に撮った命令情報
   * @param state 当該フェーズのレジスタ
   */
  private writeRecord(
    phase: CpuLogPhase,
    pending: PendingInstruction,
    state: CPURegister,
  ): void {
    const line = [
      getClockCount().toString(),
      hex4(pending.ic),
      pending.data,
      pending.name,
      phase,
      String(pending.hit),
      pending.disasm,
      formatCpuLogRegs(state),
      formatCpuLogStack(state.SP, this.readWord),
    ].join("\t");
    fs.appendFileSync(this.filePath, `${line}\n`, "utf8");
  }
}

File: retrocpu_test_framework_rs/src/mn1613/cpu_log.rs
//! テスト専用 CPU 実行ログ。
//!
//! 根拠: `asm_test_framework.mdc` §テスト専用 CPU ログ出力 /
//! `retrocpu_test_framework_ts/src/mn1613/cpu_log.ts`

use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::sync::atomic::{AtomicU64, Ordering};

use retrocpu_emu_rs::cpuboard::cpu_core::mn1613::CpuRegister;

use crate::cdb::parse_cdb;
use crate::cpu_log_mark::{
    set_active_cpu_log_marker, take_pending_cpu_log_test_name, CpuLogMarker, CpuLogTestPhase,
};
use crate::error::FrameworkError;
use crate::mn1613::disasm::Mn1613Disassembler;
use crate::types::CpuLogMode;

const STACK_WORDS: usize = 16;

fn hex4(n: u16) -> String {
    format!("{:04X}", n & 0xffff)
}

/// CDB のチェックポイントからワードアドレス → ログ用名を作る。
pub fn checkpoints_by_word_addr(cdb_text: &str) -> Result<std::collections::HashMap<u16, String>, FrameworkError> {
    let cdb = parse_cdb(cdb_text)?;
    let mut map = std::collections::HashMap::new();
    for cp in &cdb.checkpoints {
        let label = format!("{}${}", cp.name, cp.serial);
        let addr = (cp.word_addr & 0xffff) as u16;
        map.entry(addr)
            .and_modify(|prev: &mut String| {
                prev.push(',');
                prev.push_str(&label);
            })
            .or_insert(label);
    }
    Ok(map)
}

/// レジスタをログ用 1 フィールドにする。
pub fn format_cpu_log_regs(st: &CpuRegister) -> String {
    [
        format!("R0={}", hex4(st.r[0])),
        format!("R1={}", hex4(st.r[1])),
        format!("R2={}", hex4(st.r[2])),
        format!("R3={}", hex4(st.r[3])),
        format!("R4={}", hex4(st.r[4])),
        format!("SP={}", hex4(st.sp)),
        format!("STR={}", hex4(st.str)),
        format!("IC={}", hex4(st.ic)),
        format!("CSBR={:X}", st.csbr & 0xf),
        format!("SSBR={:X}", st.ssbr & 0xf),
        format!("TSR0={:X}", st.tsr0 & 0xf),
        format!("TSR1={:X}", st.tsr1 & 0xf),
        format!("OSR0={}", hex4(u16::from(st.osr[0]))),
        format!("OSR1={}", hex4(u16::from(st.osr[1]))),
        format!("OSR2={}", hex4(u16::from(st.osr[2]))),
        format!("OSR3={}", hex4(u16::from(st.osr[3]))),
        format!("NPP={:02X}", st.npp),
        format!("IISR={}", hex4(st.iisr)),
        format!("SBRB={}", hex4(st.sbrb)),
    ]
    .join(" ")
}

/// SP+1 から 16 ワードを読む。
pub fn format_cpu_log_stack(sp: u16, read_word: impl Fn(u16) -> u16) -> String {
    let mut words = Vec::with_capacity(STACK_WORDS);
    for i in 1..=STACK_WORDS {
        words.push(hex4(read_word(sp.wrapping_add(i as u16))));
    }
    words.join(" ")
}

#[derive(Debug, Clone)]
struct PendingInstruction {
    ic: u16,
    name: String,
    hit: u32,
    data: String,
    disasm: String,
}

/// テスト専用 CPU ログ。
pub struct CpuExecutionLog {
    file_path: PathBuf,
    mode: Mutex<Option<CpuLogMode>>,
    checkpoints: std::collections::HashMap<u16, String>,
    disasm: Mn1613Disassembler,
    hits: Mutex<std::collections::HashMap<u16, u32>>,
    pending: Mutex<Option<PendingInstruction>>,
    write_lock: Mutex<()>,
    log_clock: AtomicU64,
}

impl CpuExecutionLog {
    /// 出力ファイルを作る（既存なら切り詰める）。
    pub fn new(
        file_path: impl AsRef<Path>,
        cdb_text: &str,
        mode: Option<CpuLogMode>,
    ) -> Result<Arc<Self>, FrameworkError> {
        let file_path = file_path.as_ref().to_path_buf();
        if let Some(parent) = file_path.parent() {
            fs::create_dir_all(parent).map_err(|e| {
                FrameworkError::invalid_argument(format!(
                    "failed to create cpu log dir {}: {e}",
                    parent.display()
                ))
            })?;
        }
        fs::write(&file_path, "").map_err(|e| {
            FrameworkError::invalid_argument(format!(
                "failed to truncate cpu log {}: {e}",
                file_path.display()
            ))
        })?;

        let log = Arc::new(Self {
            file_path,
            mode: Mutex::new(mode),
            checkpoints: checkpoints_by_word_addr(cdb_text)?,
            disasm: Mn1613Disassembler::new(),
            hits: Mutex::new(std::collections::HashMap::new()),
            pending: Mutex::new(None),
            write_lock: Mutex::new(()),
            log_clock: AtomicU64::new(0),
        });

        set_active_cpu_log_marker(Some(log.clone()));
        if let Some(name) = take_pending_cpu_log_test_name() {
            log.append_test_mark(&name, CpuLogTestPhase::Start);
        }
        Ok(log)
    }

    /// 現在の本文モード。`None` は START/END のみ。
    pub fn log_mode(&self) -> Option<CpuLogMode> {
        *self.mode.lock().expect("mode lock")
    }

    /// 出力モードを切り替える。
    pub fn set_mode(&self, mode: Option<CpuLogMode>) {
        *self.mode.lock().expect("mode lock") = mode;
        if let Ok(mut pending) = self.pending.lock() {
            *pending = None;
        }
    }

    /// このテスト開始時点からの通過回数を捨てる（reload 時）。
    pub fn reset_hits(&self) {
        if let Ok(mut hits) = self.hits.lock() {
            hits.clear();
        }
        if let Ok(mut pending) = self.pending.lock() {
            *pending = None;
        }
    }

    /// セクション見出しを書く（本文モード無しでは書かない）。
    pub fn begin_run(&self, kind: &str, label: &str) {
        if let Ok(mut pending) = self.pending.lock() {
            *pending = None;
        }
        if self.log_mode().is_none() {
            return;
        }
        let extra = if label.is_empty() {
            String::new()
        } else {
            format!(" {label}")
        };
        let _ = self.append_line(&format!("# {kind}{extra}"));
    }

    /// フェッチ直前。
    pub fn on_before_execute(
        &self,
        state: &CpuRegister,
        read_word: impl Fn(u16) -> u16,
    ) {
        let Some(mode) = self.log_mode() else {
            if let Ok(mut pending) = self.pending.lock() {
                *pending = None;
            }
            return;
        };

        let ic = state.ic & 0xffff;
        let cp_name = self.checkpoints.get(&ic).cloned();

        match mode {
            CpuLogMode::Checkpoint => {
                let Some(name) = cp_name else {
                    if let Ok(mut pending) = self.pending.lock() {
                        *pending = None;
                    }
                    return;
                };
                let pending = self.capture_pending(ic, &name, &read_word);
                if let Ok(mut slot) = self.pending.lock() {
                    *slot = Some(pending.clone());
                }
                self.write_record("before", &pending, state, &read_word);
            }
            CpuLogMode::Instruction => {
                let pending = self.capture_pending(ic, cp_name.as_deref().unwrap_or("-"), &read_word);
                if let Ok(mut slot) = self.pending.lock() {
                    *slot = Some(pending);
                }
            }
        }
    }

    /// 命令実行直後。
    pub fn on_after_execute(
        &self,
        state: &CpuRegister,
        read_word: impl Fn(u16) -> u16,
    ) {
        if self.log_mode().is_none() {
            if let Ok(mut pending) = self.pending.lock() {
                *pending = None;
            }
            return;
        }
        let pending = self.pending.lock().ok().and_then(|mut g| g.take());
        let Some(pending) = pending else {
            return;
        };
        self.write_record("after", &pending, state, &read_word);
    }

    fn capture_pending(&self, ic: u16, name: &str, read_word: &dyn Fn(u16) -> u16) -> PendingInstruction {
        let hit = {
            let mut hits = self.hits.lock().expect("hits lock");
            let next = hits.get(&ic).copied().unwrap_or(0) + 1;
            hits.insert(ic, next);
            next
        };
        let dis = self.disasm.disassemble(ic, read_word);
        let mut data_words = Vec::new();
        for i in 0..dis.word_count {
            data_words.push(hex4(read_word(ic.wrapping_add(i as u16))));
        }
        PendingInstruction {
            ic,
            name: name.to_string(),
            hit,
            data: data_words.join(" "),
            disasm: dis.text,
        }
    }

    fn write_record(
        &self,
        phase: &str,
        pending: &PendingInstruction,
        state: &CpuRegister,
        read_word: &dyn Fn(u16) -> u16,
    ) {
        self.log_clock.fetch_add(4, Ordering::Relaxed);
        let line = [
            self.log_clock.load(Ordering::Relaxed).to_string(),
            hex4(pending.ic),
            pending.data.clone(),
            pending.name.clone(),
            phase.to_string(),
            pending.hit.to_string(),
            pending.disasm.clone(),
            format_cpu_log_regs(state),
            format_cpu_log_stack(state.sp, read_word),
        ]
        .join("\t");
        let _ = self.append_line(&line);
    }

    fn append_line(&self, line: &str) -> Result<(), FrameworkError> {
        let _guard = self.write_lock.lock().expect("write lock");
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.file_path)
            .map_err(|e| {
                FrameworkError::invalid_argument(format!(
                    "failed to open cpu log {}: {e}",
                    self.file_path.display()
                ))
            })?;
        writeln!(file, "{line}").map_err(|e| {
            FrameworkError::invalid_argument(format!(
                "failed to write cpu log {}: {e}",
                self.file_path.display()
            ))
        })
    }
}

impl CpuLogMarker for CpuExecutionLog {
    fn append_test_mark(&self, name: &str, phase: CpuLogTestPhase) {
        let p = match phase {
            CpuLogTestPhase::Start => "START",
            CpuLogTestPhase::End => "END",
        };
        let _ = self.append_line(&format!("{name} {p}"));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::CpuLogMode;

    #[test]
    fn format_regs_and_stack() {
        let st = CpuRegister {
            r: [3, 4, 0, 0, 0],
            sp: 0xfffe,
            ..Default::default()
        };
        assert!(format_cpu_log_regs(&st).contains("R0=0003"));
        let stack = format_cpu_log_stack(0xfffe, |a| a);
        assert_eq!(stack.split(' ').count(), 16);
    }

    #[test]
    fn checkpoint_mode_writes_before_after() -> Result<(), FrameworkError> {
        let dir = tempfile::tempdir().expect("tempdir");
        let log_path = dir.path().join("cpu.log");
        let cdb = "L:__CP$add_enter$0001:0404\n";
        let log = CpuExecutionLog::new(
            &log_path,
            cdb,
            Some(CpuLogMode::Checkpoint),
        )?;
        log.begin_run("call", "gl_add");
        let read = |a: u16| if a == 0x0202 { 0x5809 } else { 0 };
        let before = CpuRegister {
            ic: 0x0202,
            r: [3, 4, 0, 0, 0],
            sp: 0xfffe,
            ..Default::default()
        };
        log.on_before_execute(&before, read);
        let after = CpuRegister {
            ic: 0x0203,
            r: [7, 4, 0, 0, 0],
            sp: 0xfffe,
            ..Default::default()
        };
        log.on_after_execute(&after, read);
        let text = fs::read_to_string(&log_path).expect("read log");
        assert!(text.contains("add_enter$0001\tbefore\t1\tA R0, R1"));
        assert!(text.contains("add_enter$0001\tafter\t1"));
        Ok(())
    }
}
